File: leetcode/go/169.majority-element.go
/*
 * @lc app=leetcode id=169 lang=golang
 *
 * [169] Majority Element
 */

// @lc code=start
func majorityElement(nums []int) int {
	ret := nums[0]
	length := len(nums)
	number := 1
	for i := 1; i < length; i++ {
		if number == 0 {
			ret = nums[i]
			number++
			continue
		}
		if nums[i] == ret {
			number++
		} else {
			number--
		}
	}
	return ret
}

// @lc code=end

File: leetcode/cpp/169.majority-element.cpp
/*
 * @lc app=leetcode id=169 lang=cpp
 *
 * [169] Majority Element
 */
#include <vector>
#include <iostream>
using namespace std;
// @lc code=start
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int val = nums[0];
        int count = 1;
        for (int i = 1; i < nums.size(); ++i){
            if(count == 0){
                val = nums[i];
                 count = 1;
            } else {
                if(nums[i] == val){
                    ++count;
                } else {
                    --count;
                }
            }
        }
        return val;
    }
};
// @lc code=end

